core/core.go

```diff
@@ -214,30 +214,39 @@ func (n *IpfsNode) StartOnlineServices(ctx context.Context) error {
 		return err
 	}
 
-	peerhost, err := constructPeerHost(ctx, n.Repo.Config(), n.Identity, n.Peerstore)
+	peerhost, err := constructPeerHost(ctx, n.Identity, n.Peerstore)
 	if err != nil {
 		return debugerror.Wrap(err)
 	}
 	n.PeerHost = peerhost
 
-	// setup diagnostics service
-	n.Diagnostics = diag.NewDiagnostics(n.Identity, n.PeerHost)
+	// this block is the set of services which need to be initialized with the host
+	// and _before_ we start listening.
+	{
+		// setup diagnostics service
+		n.Diagnostics = diag.NewDiagnostics(n.Identity, n.PeerHost)
 
-	// setup routing service
-	dhtRouting, err := constructDHTRouting(ctx, n.PeerHost, n.Repo.Datastore())
-	if err != nil {
-		return debugerror.Wrap(err)
-	}
-	n.Routing = dhtRouting
+		// setup routing service
+		dhtRouting, err := constructDHTRouting(ctx, n.PeerHost, n.Repo.Datastore())
+		if err != nil {
+			return debugerror.Wrap(err)
+		}
+		n.Routing = dhtRouting
 
-	// setup exchange service
-	const alwaysSendToPeer = true // use YesManStrategy
-	bitswapNetwork := bsnet.NewFromIpfsHost(n.PeerHost, n.Routing)
-	n.Exchange = bitswap.New(ctx, n.Identity, bitswapNetwork, n.Blockstore, alwaysSendToPeer)
+		// setup exchange service
+		const alwaysSendToPeer = true // use YesManStrategy
+		bitswapNetwork := bsnet.NewFromIpfsHost(n.PeerHost, n.Routing)
+		n.Exchange = bitswap.New(ctx, n.Identity, bitswapNetwork, n.Blockstore, alwaysSendToPeer)
 
-	// setup name system
-	// TODO implement an offline namesys that serves only local names.
-	n.Namesys = namesys.NewNameSystem(n.Routing)
+		// setup name system
+		// TODO implement an offline namesys that serves only local names.
```

Comment:
I know you didnt add this TODO here, but its worth noting that the offline routing system `routing/offline` will do exactly this